sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/cpd/internal/DefaultCpdTokens.java

```diff
@@ -39,18 +43,39 @@
   private int currentIndex = 0;
   private StringBuilder sb = new StringBuilder();
   private TextRange lastRange;
+  private boolean excluded;
 
-  public DefaultCpdTokens(SensorStorage storage) {
+  public DefaultCpdTokens(Settings settings, SensorStorage storage) {
     super(storage);
+    this.settings = settings;
   }
 
   @Override
   public DefaultCpdTokens onFile(InputFile inputFile) {
     Preconditions.checkNotNull(inputFile, "file can't be null");
     this.inputFile = (DefaultInputFile) inputFile;
+    String language = inputFile.language();
+    if (language != null && isSkipped(language)) {
+      this.excluded = true;
+    } else {
+      String[] cpdExclusions = settings.getStringArray(CoreProperties.CPD_EXCLUSIONS);
+      for (PathPattern cpdExclusion : PathPattern.create(cpdExclusions)) {
+        if (cpdExclusion.match(inputFile)) {
+          this.excluded = true;
+        }
+      }
+    }
     return this;
   }
 
+  boolean isSkipped(String language) {
+    String key = "sonar.cpd." + language + ".skip";
+    if (settings.hasKey(key)) {
+      return settings.getBoolean(key);
+    }
+    return settings.getBoolean(CoreProperties.CPD_SKIP_PROPERTY);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this use of "CPD_SKIP_PROPERTY"; it is deprecated. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)